--- old_ubi.h	2022-06-25 01:18:56.845298385 +0800
+++ ubi.h	2022-06-25 09:36:59.024270822 +0800
@@ -27,6 +27,7 @@
 #include <linux/mtd/mtd.h>
 #include <linux/mtd/ubi.h>
 #include <linux/pgtable.h>
+#include <linux/semaphore.h>
 
 #include "ubi-media.h"
 
@@ -628,6 +629,12 @@
 	int thread_enabled;
 	char bgt_name[sizeof(UBI_BGT_NAME_PATTERN)+2];
 
+	/* async scan's stuff */
+	struct semaphore peb_sem;
+	spinlock_t pnum_lock;
+	spinlock_t image_seq_lock;
+	spinlock_t ai_lock;
+
 	/* I/O sub-system's stuff */
 	long long flash_size;
 	int peb_count;
@@ -785,6 +792,8 @@
 	struct kmem_cache *aeb_slab_cache;
 	struct ubi_ec_hdr *ech;
 	struct ubi_vid_io_buf *vidb;
+	/* async scan's stuff */
+	spinlock_t snum_lock;
 };
 
 /**
