--- drivers/mtd/ubi/ubi.bak	2022-05-07 00:23:49.069617970 +0800
+++ drivers/mtd/ubi/ubi.h	2022-05-07 22:16:28.814118638 +0800
@@ -27,6 +27,8 @@
 #include <linux/mtd/mtd.h>
 #include <linux/mtd/ubi.h>
 #include <linux/pgtable.h>
+#include <linux/kthread.h>
+#include <linux/semaphore.h>
 
 #include "ubi-media.h"
 
@@ -628,6 +630,12 @@
 	int thread_enabled;
 	char bgt_name[sizeof(UBI_BGT_NAME_PATTERN)+2];
 
+	/* async scan's stuff */
+	struct semaphore peb_sem;
+	struct mutex pnum_mutex;
+	struct mutex image_seq_mutex;
+	struct mutex ai_mutex;
+
 	/* I/O sub-system's stuff */
 	long long flash_size;
 	int peb_count;
@@ -785,6 +793,8 @@
 	struct kmem_cache *aeb_slab_cache;
 	struct ubi_ec_hdr *ech;
 	struct ubi_vid_io_buf *vidb;
+	/* async scan's stuff */
+	int err;
 };
 
 /**
